services/utilServices/AuthService.js

```diff
@@ -0,0 +1,96 @@
+const axios = require("axios")
+const queryString = require("query-string")
+const uuid = require("uuid/v4")
+
+// Import error
+const { AuthError } = require("@errors/AuthError")
+const { ForbiddenError } = require("@errors/ForbiddenError")
+
+const validateStatus = require("@utils/axios-utils")
+const jwtUtils = require("@utils/jwt-utils")
+
+const { CLIENT_ID, CLIENT_SECRET, REDIRECT_URI } = process.env
+
+class AuthService {
+  async getAuthRedirectDetails() {
+    const state = uuid()
+    const githubAuthUrl = `https://github.com/login/oauth/authorize?client_id=${CLIENT_ID}&state=${state}&scope=repo`
+
+    const token = jwtUtils.signToken({ state })
+
+    return { redirectUrl: githubAuthUrl, cookieToken: token }
+  }
+
+  async getGithubAuthToken({ csrfState, code, state }) {
+    const isCsrfValid =
+      jwtUtils.verifyToken(csrfState) &&
+      jwtUtils.decodeToken(csrfState).state === state
+    if (!isCsrfValid) throw new ForbiddenError()
+
+    const params = {
+      code,
+      redirect_uri: REDIRECT_URI,
+      state,
+    }
+
+    const resp = await axios.post(
+      "https://github.com/login/oauth/access_token",
+      params,
+      {
+        auth: {
+          username: CLIENT_ID,
+          password: CLIENT_SECRET,
+        },
+      }
+    )
+
+    const { access_token: accessToken } = queryString.parse(resp.data)
+    if (!accessToken) throw new AuthError("Access token not found")
+
+    // Retrieve user information to put into access token
+    const endpoint = `https://api.github.com/user`
+    const userResp = await axios.get(endpoint, {
+      validateStatus,
+      headers: {
+        Authorization: `token ${accessToken}`,
+        Accept: "application/vnd.github.v3+json",
+        "Content-Type": "application/json",
+      },
+    })
+
+    let userId
+    if (userResp.data) userId = userResp.data.login
```

Comment:
```const userId = userResp.data && userResp.data.login```

Reply:
This would result in a `false` value for userId though, which I think we want to avoid

Reply:
I was thinking that signtoken might have unexpected behaviour - but it seems fine, https://github.com/isomerpages/isomercms-backend/pull/328/commits/bbd8028e94e4cd88b5e9969fc1903f2184fb0df5